src/python/WMCore/Services/DBS/DBS3Reader.py

```diff
@@ -883,3 +885,167 @@ def listDatasetParents(self, childDataset):
             msg += "DBSReader.listDatasetParents(%s)\n" % childDataset
             msg += "%s\n" % formatEx3(ex)
             raise DBSReaderError(msg)
+
+    def getParentFilesGivenParentDataset(self, parentDataset, childLFN):
```

Comment:
This function should be easily parallelizeable, but it requires complete change of its logic and tools. I'll provide here a recipe and pointers and you can look/adjust the code. First, you need pycurl_manager getdata function
```
from WMCore.Services.pycurl_manager import getdata as multi_getdata
```
Then you need to construct a list of urls you're going to concurrently send to DBS, e.g.
```
uri = /dbs_uri # replace with main DBS URI
urls = ["%s/parents?run_num=%s&lumi_list=%s" % (uri, f['run_num'], f['lumi_section_num']) for f in fInfo]
```
then you can implement similar logic as shown here:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/MicroService/Unified/RequestInfo.py#L150
For example:
```
data = multi_getdata(urls, ckey(), cert())
parentFiles = []
for row in data:
     # request url
     req = row['url'].split('/')[-1]
     # response data
     data = json.loads(row['data'])
     # parse data here and extract file name from response
     parentFiles.append( data['logical_file_name'] ) # replace logical_file_name with whatever key DBS uses in response
```
So, to sum-up.
1. you will not use dbs client APIs since they do not support concurrency
2. you construct list of all urls you want to fetch at once
3. multi_getdata will do proper thing: it will place all calls concurrently and do not saturate DBS server (internally it split input list of urls in to a chunk (default size is 100) and send the chunks).
4. you'll collect results from pycurl and extract responses and construct your output.

Reply:
@bbockelm , Brian, if I'm not mistaken requests library only provides API to fetch a single request from the server. Therefore to implement concurrency we either need to write asyncio code (but it is only available in python 3.5+) or write custom code via thread pool, etc. Therefore it till take some effort/code which we'll need to mantain. This is what we're doing with pycurl already. Plus pycurl basically calls curl C library, so it's is fast. How requests libraryunderlying logic is implemented I don't know. Seems like a good student project to me to extend requests library and implement concurrent APIs and perform benchmarks wrt pycurl.

Reply:
Sounds good, thanks for the update!